cpmpy/transformations/safening.py

```diff
@@ -0,0 +1,239 @@
+        :param: safen toplevel: list of expression types that need to be safened, even when toplevel. Used when
+                                 a solver does not support unsafe values in it's API (e.g., OR-Tools for `div`).
+    """
+
+    if _toplevel is None:
+        assert _nbc is None, f"_nbc is an internal argument, should not be filled by caller but got {_nbc}"
+        toplevel_call = True
+        _toplevel = []
+        _nbc = _toplevel # at the toplevel of the contraint model, the neirest Boolean context is just toplevel
+    else:
+        toplevel_call = False
+        assert isinstance(_toplevel, list), f"_toplevel argument must be of type list but got {type(_toplevel)}"
+        assert isinstance(_nbc, list), f"_nbc argument must be of type list but got {type(_nbc)}"
+
+    new_lst = []
+    for cpm_expr in lst_of_expr:
+
+        if is_num(cpm_expr) or isinstance(cpm_expr, _NumVarImpl):
+            new_lst.append(cpm_expr)
+
+        elif isinstance(cpm_expr, (list,tuple)):
+            new_lst.append(no_partial_functions(cpm_expr, _toplevel, _nbc, safen_toplevel))
+
+        elif isinstance(cpm_expr, NDVarArray):  # TODO efficiency: and cpm_expr.has_subexpr()
+            # efficiency: flatten into single iterator, then reshape to n-dimensional again
+            new_cpm_expr = cpm_array(no_partial_functions(cpm_expr.flat, _toplevel, _nbc, safen_toplevel)).reshape(cpm_expr.shape)
+            new_lst.append(new_cpm_expr)
+
+        elif isinstance(cpm_expr, DirectConstraint):  # do not recurse into args
+            new_lst.append(cpm_expr)
+
+        else:
+            assert isinstance(cpm_expr, Expression), f"each `cpm_expr` should be an Expression at this point, not {type(cpm_expr)}"
+            args = cpm_expr.args
+
+            if cpm_expr.is_bool() and toplevel_call is False:  # a Boolean context, create a new _nbc
+                # if we are currently toplevel, no need to create a new nbc, will append to toplevel anyway
+                _nbc = []
+
+            # recurse over the arguments of the expression
+            # TODO efficiency: if cpm_expr.has_subexpr()
+            new_args = no_partial_functions(args, _toplevel, _nbc, safen_toplevel=safen_toplevel)
+            if any((a1 is not a2) for a1,a2 in zip(new_args,args)):  # efficiency (hopefully): only copy if an arg changed
+                cpm_expr = copy(cpm_expr)
+                cpm_expr.update_args(new_args)
+                args = new_args
+
+            if cpm_expr.is_bool() and len(_nbc) != 0 and toplevel_call is False:
+                # a nested Boolean context, conjoin my Boolean expression with _nbc
+                # in `b <-> (Arr[idx] == 5)`, this would trigger for cpm_expr (Arr[idx] == 5)
+                cpm_expr = cpm_all(_nbc) & cpm_expr
+
+            elif isinstance(cpm_expr, GlobalFunction) and cpm_expr.name == "element":
+                if _nbc is _toplevel and cpm_expr.name not in safen_toplevel: # no need to safen
```

Comment:
I do not get this, think it leads to a mistake where nested element does not get safened: `(a / arr[idx]) == 0` (added a testcase)

Reply:
or even just (a * arr[idx]) == 0 does not get safened

Reply:
ah i see it gets flattened to toplevel anyway? don't know if I like relying on flattening for getting things 'safe' on toplevel

Reply:
and solvers are using strict semantics on toplevel, while we want to use relational semantics in the example I gave?

Reply:
Hmm no, in the example you gave the neirest Boolean context is "toplevel" so, the constraint `a * arr[idx] == 0` can only be satisfied if `idx` is within the bounds of the array.
There is no real diffrence between `a * arr[idx] == 0` and `arr[idx] == 0` from a safening point of view.